File: backend/internal/biz/system/admin/base_table_source.go
package biz

import (
	"context"
	"fmt"
	"sort"

	adminv1 "github.com/liujitcn/kratos-admin/backend/api/gen/go/system/admin/v1"
	"github.com/liujitcn/kratos-admin/backend/internal/biz/system/admin/dto"
	commonv1 "github.com/liujitcn/kratos-core/api/gen/go/common/v1"
	"github.com/liujitcn/kratos-core/biz"
	"github.com/liujitcn/kratos-core/errorsx"
	"github.com/liujitcn/kratos-kit/database/gorm"
)

// BaseTableSourceCase 提供数据备份、数据归档和代码生成共用的数据源元数据。
type BaseTableSourceCase struct {
	*biz.BaseCase
}

// NewBaseTableSourceCase 创建数据源元数据业务实例。
func NewBaseTableSourceCase(baseCase *biz.BaseCase) *BaseTableSourceCase {
	return &BaseTableSourceCase{BaseCase: baseCase}
}

// OptionBaseTableSource 查询已初始化的数据源名称。
func (c *BaseTableSourceCase) OptionBaseTableSource(context.Context) (*commonv1.StringValues, error) {
	names := make([]string, 0, len(c.GormClients))
	for name, client := range c.GormClients {
		if name == "" || client == nil || client.DB == nil {
			continue
		}
		names = append(names, name)
	}
	sort.Strings(names)
	return &commonv1.StringValues{Value: names}, nil
}

// OptionBaseTable 查询指定数据源中的数据库表选项。
func (c *BaseTableSourceCase) OptionBaseTable(ctx context.Context, req *adminv1.OptionBaseTableRequest) (*adminv1.OptionBaseTableResponse, error) {
	client, err := GormClientBySourceName(c.BaseCase, req.GetSourceName())
	if err != nil {
		return nil, errorsx.InvalidArgument("请选择已初始化的数据源").WithCause(err)
	}
	var tables []dto.CodeGenDatabaseTable
	tables, err = listDatabaseTableMetadata(ctx, client, nil)
	if err != nil {
		return nil, fmt.Errorf("查询数据源 %s 的数据库表失败: %w", req.GetSourceName(), err)
	}
	options := make([]*adminv1.BaseTableOption, 0, len(tables))
	for _, table := range tables {
		options = append(options, &adminv1.BaseTableOption{Name: table.TableName, Comment: table.TableComment})
	}
	return &adminv1.OptionBaseTableResponse{Tables: options}, nil
}

// listDatabaseTableMetadata 查询指定客户端的数据表名和表描述。
func listDatabaseTableMetadata(ctx context.Context, database *gorm.Client, tableNames []string) ([]dto.CodeGenDatabaseTable, error) {
	query := database.DB.WithContext(ctx).
		Table("information_schema.tables").
		Select("table_name, table_comment").
		Where("table_schema = DATABASE()").
		Where("table_type = ?", "BASE TABLE")
	if len(tableNames) > 0 {
		query = query.Where("table_name IN ?", tableNames)
	}
	var tableInfos []dto.CodeGenDatabaseTable
	err := query.Order("table_name").Find(&tableInfos).Error
	return tableInfos, err
}

// GormClientBySourceName 按数据源名称获取已初始化的 GORM 客户端。
func GormClientBySourceName(baseCase *biz.BaseCase, sourceName string) (*gorm.Client, error) {
	if sourceName == "" {
		return nil, fmt.Errorf("数据源名称不能为空")
	}
	client := baseCase.GormClients[sourceName]
	if client == nil || client.DB == nil {
		return nil, fmt.Errorf("数据源 %s 未初始化", sourceName)
	}
	return client, nil
}

// sourceTableKey 返回数据源和数据表组成的稳定索引键。
func sourceTableKey(sourceName, tableName string) string {
	return sourceName + "\x00" + tableName
}

File: frontend/admin/packages/modules/system/src/api/system/admin/v1/base_table_source.ts
import service from "@liujitcn/kratos-admin-core/request";
import type { StringValues } from "@liujitcn/kratos-admin-system/rpc/common/v1/types";
import type {
  BaseTableSourceService,
  OptionBaseTableRequest,
  OptionBaseTableResponse,
  OptionBaseTableSourceRequest
} from "@liujitcn/kratos-admin-system/rpc/system/admin/v1/base_table_source";

const BASE_TABLE_SOURCE_URL = "/v1/admin/base/table-source";

/** 数据源元数据服务实现。 */
export class BaseTableSourceServiceImpl implements BaseTableSourceService {
  /** 查询已初始化的数据源名称。 */
  OptionBaseTableSource(request: OptionBaseTableSourceRequest): Promise<StringValues> {
    return service<OptionBaseTableSourceRequest, StringValues>({
      url: `${BASE_TABLE_SOURCE_URL}/option`,
      method: "get",
      params: request
    });
  }

  /** 查询指定数据源中的数据库表选项。 */
  OptionBaseTable(request: OptionBaseTableRequest): Promise<OptionBaseTableResponse> {
    return service<OptionBaseTableRequest, OptionBaseTableResponse>({
      url: `${BASE_TABLE_SOURCE_URL}/table/option`,
      method: "get",
      params: request
    });
  }
}

/** 数据源元数据服务实例。 */
export const defBaseTableSourceService = new BaseTableSourceServiceImpl();
